modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -347,6 +373,82 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
 
         self._info("Training complete!")
 
+    def prepare_data(self, batch: tuple) -> tuple[list, torch.Tensor, Union[torch.Tensor, dict]]:
+        sample_ids = batch[0]
+        if isinstance(sample_ids, torch.Tensor):
+            sample_ids = sample_ids.tolist()
+        elif isinstance(sample_ids, tuple):
+            sample_ids = list(sample_ids)
+
+        assert isinstance(sample_ids, list), "Cannot parse result from DataLoader"
+
+        if self._label_tranformer_function is None:
+            target = batch[2].to(self._device)
+        else:
+            target = self._label_tranformer_function(batch[2]).to(self._device)
+
+        data: Union[torch.Tensor, dict]
+        if isinstance(batch[1], torch.Tensor):
+            data = batch[1].to(self._device)
+        elif isinstance(batch[1], dict):
+            data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
+            for name, tensor in batch[1].items():
+                data[name] = tensor.to(self._device)
+        else:
+            raise ValueError(
+                "The format of the data provided is not supported in modyn. "
+                "Please use either torch tensors or dict[str, torch.Tensor]"
+            )
+
+        return sample_ids, target, data
+
+    def sample_data(self) -> None:
```

Comment:
could we rename this to `downsample_trigger_training_set`? or something else if you have a better idea? basically, sample_data sounds like it is a very general function that would also be used in BtS, however, this is the StB impl, right?

Reply:
Yep, renamed. Probably in the future there will be the per_class implementation too